src/cuda/api/memory.hpp

```diff
@@ -221,6 +222,117 @@ inline void copy(void *destination, const void *source, size_t num_bytes)
 	}
 }
 
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A CUDA array @ref cuda::array::array_t
+ * @param source A pointer to a a memory region of size `destination.size() * sizeof(T)`
+ */
+template<typename T>
+void copy(const array::array_t<T, 3>& destination, const void *source)
+{
+	cudaMemcpy3DParms copyParams = {0};
+
+	// make_cudaPitchedPtr expects a void* pointer
+	void* source_ = const_cast<void*>(source);
+
+	// How to create a pitched ptr from a `const void*` ?
+	copyParams.srcPtr = make_cudaPitchedPtr(
+	    source_, destination.dims()[0] * sizeof(T),  destination.dims()[0], destination.dims()[1]);
+
+	copyParams.dstArray = destination.get();
+
+	cudaExtent ext = make_cudaExtent(destination.dims()[0], destination.dims()[1], destination.dims()[2]);
+
+	copyParams.extent = ext;
+
+	copyParams.kind = cudaMemcpyDefault;
+
+	auto result = cudaMemcpy3D(&copyParams);
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying into array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A pointer to a a memory region of size `source.size() * sizeof(T)`
```

Comment:
Explain what memory spaces the pointer can point to. Also, clarify what you mean by the "size" of the memory region.